tests/Neo.UnitTests/SmartContract/Native/UT_CryptoLib.cs

```diff
@@ -427,5 +436,414 @@ public void TestKeccak256_BlankString()
             // Assert
             Assert.AreEqual(expectedHashHex, outputHashHex, "Keccak256 hash did not match expected value for blank string.");
         }
+
+        [TestMethod]
+        // TestVerifyWithECDsa_CustomTxWitness_SingleSig builds custom witness verification script for single Koblitz public key
+        // and ensures witness check is passed for the following message:
+        //
+        //	keccak256([4-bytes-network-magic-LE, txHash-bytes-BE])
+        //
+        // The proposed witness verification script has 110 bytes length, verification costs 2154270 GAS including Invocation script execution.
+        // The user has to sign the keccak256([4-bytes-network-magic-LE, txHash-bytes-BE]).
+        public void TestVerifyWithECDsa_CustomTxWitness_SingleSig()
+        {
+            byte[] privkey = "7177f0d04c79fa0b8c91fe90c1cf1d44772d1fba6e5eb9b281a22cd3aafb51fe".HexToBytes();
+            ECPoint pubKey = ECPoint.Parse("04fd0a8c1ce5ae5570fdd46e7599c16b175bf0ebdfe9c178f1ab848fb16dac74a5d301b0534c7bcf1b3760881f0c420d17084907edd771e1c9c8e941bbf6ff9108", ECCurve.Secp256k1);
+
+            // vrf is a builder of witness verification script corresponding to the public key.
+            using ScriptBuilder vrf = new();
+            vrf.EmitPush((byte)NamedCurveHash.secp256k1Keccak256); // push Koblitz curve identifier and Keccak256 hasher.
+            vrf.Emit(OpCode.SWAP); // swap curve identifier with the signature.
+            vrf.EmitPush(pubKey.EncodePoint(true)); // emit the caller's public key.
+
+            // Construct and push the signed message. The signed message is effectively the network-dependent transaction hash,
+            // i.e. msg = [4-network-magic-bytes-LE, tx-hash-BE]
+            // Firstly, retrieve network magic (it's uint32 wrapped into BigInteger and represented as Integer stackitem on stack).
+            vrf.EmitSysCall(ApplicationEngine.System_Runtime_GetNetwork); // push network magic (Integer stackitem), can have 0-5 bytes length serialized.
+
+            // Convert network magic to 4-bytes-length LE byte array representation.
+            vrf.EmitPush(0x100000000); // push 0x100000000.
+            vrf.Emit(OpCode.ADD, // the result is some new number that is 5 bytes at least when serialized, but first 4 bytes are intact network value (LE).
+                    OpCode.PUSH4, OpCode.LEFT); // cut the first 4 bytes out of a number that is at least 5 bytes long, the result is 4-bytes-length LE network representation.
```

Comment:
@neo-project/core I think it's time for us to decide: we should move this witness builder code to some method in the core, so that users can just call this method like standard `Contract.CreateMultiSigRedeemScript` and `Contract.CreateSignatureRedeemScript`.

I'd suggest to move it to `Contract.CreateKoblitzMultiSigRedeemScript` and `Contract.CreateKoblitzSignatureRedeemScript` or something like this. What do you think?

Reply:
Agree

Reply:
Discussed in Discored, we're not adding this code to the core.